pkg/compactor/compactor.go

```diff
@@ -383,42 +392,39 @@ func (c *MultitenantCompactor) starting(ctx context.Context) error {
 	c.bucketClient = bucketindex.BucketWithGlobalMarkers(c.bucketClient)
 
 	// Initialize the compactors ring if sharding is enabled.
-	lifecyclerCfg := c.compactorCfg.ShardingRing.ToLifecyclerConfig()
-	c.ringLifecycler, err = ring.NewLifecycler(lifecyclerCfg, ring.NewNoopFlushTransferer(), "compactor", CompactorRingKey, false, c.logger, prometheus.WrapRegistererWithPrefix("cortex_", c.registerer))
+	c.ring, c.ringLifecycler, err = newRingAndLifecycler(c.compactorCfg.ShardingRing, c.logger, c.registerer)
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring lifecycler")
+		return err
 	}
 
-	c.ring, err = ring.New(lifecyclerCfg.RingConfig, "compactor", CompactorRingKey, c.logger, prometheus.WrapRegistererWithPrefix("cortex_", c.registerer))
+	c.ringSubservices, err = services.NewManager(c.ringLifecycler, c.ring)
 	if err != nil {
-		return errors.Wrap(err, "unable to initialize compactor ring")
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	c.ringSubservices, err = services.NewManager(c.ringLifecycler, c.ring)
-	if err == nil {
-		c.ringSubservicesWatcher = services.NewFailureWatcher()
-		c.ringSubservicesWatcher.WatchManager(c.ringSubservices)
-
-		err = services.StartManagerAndAwaitHealthy(ctx, c.ringSubservices)
+	c.ringSubservicesWatcher = services.NewFailureWatcher()
+	c.ringSubservicesWatcher.WatchManager(c.ringSubservices)
+	if err = c.ringSubservices.StartAsync(ctx); err != nil {
+		return errors.Wrap(err, "unable to start compactor ring dependencies")
 	}
 
-	if err != nil {
+	ctxTimeout, cancel = context.WithTimeout(ctx, c.compactorCfg.ShardingRing.WaitActiveInstanceTimeout)
+	defer cancel()
+	if err = c.ringSubservices.AwaitHealthy(ctxTimeout); err != nil {
```

Comment:
I'm not sure if this is the right choice. There was previously no timeout on how long we'd wait for the ring and lifecycler to become healthy. This was fine because the `ring.Lifecycler` doesn't actually have a "start" method and so it isn't doing anything when starting and becoming healthy. However, because we're now using the `ring.BasicLifecycler`, it actually joins the ring when starting and so it [might hit this issue](https://github.com/cortexproject/cortex/issues/3914) if there was no timeout for startup.

Reply:
I think applying `WaitActiveInstanceTimeout` here as well is correct.